### DISCUSSION TITLE
how to run migrations of libraries?

### DISCUSSION THREAD

**Discussion by nohkumado** [diesel-rs/diesel#3756] [Q&A] [answered]
Hello !
If there is allready a guide i didn't find it? 
I have split my project into a library with my common DB Tasks, and derive from that other rust binary crates that expand the DB with additional tables....
The library works, but i am stuck at how to let diesel call from the binary crate the migrations from the library crate (using the DB from the binary)?
In my cargo.toml i specified the path to the library, running ````diesel migration run```` only installs the tables in the binary, no cascading over the library ? At least i didn't find it?

**Comment by weiznich** [comment]
If you want to run your migrations via diesel-cli, you need to execute `diesel migration run` in **both** the library and the binary. It's expected that only the local migration directory is applied.

**Reply by nohkumado** [reply]
Hmm ok..... Problem is the database settings are different for each project... If i symlink the files of the library into the binary crate? Or is there a better way?

**Reply by weiznich** [reply, marked as answer]
Whether or not there is a better way really depends on your specific needs. These are not really clear from the limited amount of information provided.

Anyway: Diesel CLI currently expects to apply a migration either in the current folder or specified via the `--mirgatio-dir` flag. There is no built-in or otherwise supported way to combine several directories there.

**Reply by nohkumado** [reply]
ok! that worked indeed! would be nice to have a sort of #include <../mylib/migrations>  instead of having to run migrations twice (or more if i have to split  the project further into specific subprojects with a common funcitonality....)

**Reply by weiznich** [reply]
I'm happy to discuss a well written proposal for such a feature. Please mind that this would need to include a specification how this is supposed to work and at least some idea on how this could be implemented.

**Comment by nohkumado** [comment]
:D i understand
Is there some docu on how the actual migration system us working?
Unfortunately i am far from firm in rust to be able to read the code
easily....

Ciao

Bruno Böttcher
--
35 rue de la république, FR-6720 Schwindratzheim
email: ***@***.***, ***@***.***
Fon:+33 3 88 89 91, Mob:+33 6 76 55 82 68
-------------------------------------------------
Dev: Java/Perl/PHP OS:GNU/LINUX, Android

Georg Semmler ***@***.***> schrieb am Mi., 23. Aug. 2023,
18:00:

**Comment by nohkumado** [comment]
do i understand correctly that you store in  __diesel_schema_migrations the
numerical prefix of the file in the migrations directory after having
performed something like a psql -f <the_file>?
probably preceded by a sort of list-dir function that alphabetically lists
the available files?

if i may suggest, adding a column to that table with the content of the
migration-dir argument would already solve part of, at least my, problems:
atm, when reverting i have to know myself from where came the update, with
the path available the revert would know where to fetch the down.sql from?
i could live then with running by hand the migrations from either directory
then :D
ideally though a mixed list taken from all used migration-dirs and sorted
by the numerical tag would IMHO be the solution of choice to apply the
sub-lib migrations?
if that seems acceptable adding an include directive to add migration-dirs
inside the up and down files should be straightforward?

BTW is my use case that exotic? just wondering....



Am Do., 24. Aug. 2023 um 13:06 Uhr schrieb Noh Kumado ***@***.***>:

> :D i understand
> Is there some docu on how the actual migration system us working?
> Unfortunately i am far from firm in rust to be able to read the code
> easily....
>
> Ciao
>
> Bruno Böttcher
> --
> 35 rue de la république, FR-6720 Schwindratzheim
> email: ***@***.***, ***@***.***
> Fon:+33 3 88 89 91, Mob:+33 6 76 55 82 68
> -------------------------------------------------
> Dev: Java/Perl/PHP OS:GNU/LINUX, Android
>
> Georg Semmler ***@***.***> schrieb am Mi., 23. Aug. 2023,
> 18:00:
>
>> I'm happy to discuss a well written proposal for such a feature. Please
>> mind that this would need to include a specification how this is supposed
>> to work and at least some idea on how this could be implemented.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/diesel-rs/diesel/discussions/3756#discussioncomment-6804603>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ADR5CZ5WSVM34PSAZHEWOT3XWYSLLANCNFSM6AAAAAA3VWWVFY>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

-- 
Bruno Böttcher
--
35 rue de la république, FR-6720 Schwindratzheim
email: ***@***.***, ***@***.***
Fon:+33 3 88 89 91, Mob:+33 6 76 55 82 68
-------------------------------------------------
Dev: Java/Perl/PHP OS:GNU/LINUX, Android

**Reply by weiznich** [reply]
Such a change is unfortunately not possible for several reasons. First of all it would break all existing database instances that already have some migrations applied, as we just cannot fill in the path there. Additionally it would break the usage of diesel-cli in several systems at once as the migration-dir stored in the database wouldn't necessarily match the actual migration path for all local development copies of that directory for all developers.

**Comment by nohkumado** [comment]
ok! no problem, will do my own shell scripts then!

Am Fr., 25. Aug. 2023 um 11:55 Uhr schrieb Georg Semmler <
***@***.***>:

> Such a change is unfortunately not possible for several reasons. First of
> all it would break all existing database instances that already have some
> migrations applied, as we just cannot fill in the path there. Additionally
> it would break the usage of diesel-cli in several systems at once as the
> migration-dir stored in the database wouldn't necessarily match the actual
> migration path for all local development copies of that directory for all
> developers.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/diesel-rs/diesel/discussions/3756#discussioncomment-6821584>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR5CZZZDONXEEJ3UR477U3XXBY7RANCNFSM6AAAAAA3VWWVFY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Bruno Böttcher
--
35 rue de la république, FR-6720 Schwindratzheim
email: ***@***.***, ***@***.***
Fon:+33 3 88 89 91, Mob:+33 6 76 55 82 68
-------------------------------------------------
Dev: Java/Perl/PHP OS:GNU/LINUX, Android
